### PR TITLE
Site Editor: Add 'area' selection to convert to template part flow.

### LINKED ISSUE #29222: Create template part flow: Add "Area" selection input to modal.
When manually creating a template part from a selection of blocks, the following modal appears:

Along with the template part name, we should also provide an affordance to specify the template part `area` (Header, Footer, etc).

We might use the dedicated icons for template parts in a segmented control arrangement to achieve this.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
## Description

Attempts to resolve https://github.com/WordPress/gutenberg/issues/29222 - by adding selection for 'area' in this modal per design suggestion https://github.com/WordPress/gutenberg/issues/29222#issuecomment-791381586 (update to https://github.com/WordPress/gutenberg/pull/30395#issuecomment-810897861)

## How has this been tested?

To Test:
1. Load the Site Editor (must have a block based theme such as tt1-blocks activated).
2. Select an/some existing blocks in the editor (or create new ones).
3. On the block toolbar, click the ellipsis menu and select the "Convert to Template Part" option.
4. A modal should appear to name the template part and select its area (this area selection is what has been added in this PR).  
5. Create the template part and verify its block variation corresponds to the area selected in step 4.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by Addison-Stavlo** [comment]
cc @jameskoster - how do you feel about the styles and interaction at this point?

**Comment by jameskoster** [comment]
Everything works as expected :) 

I made a couple of adjustments to the design to better match other similar UI elements. Mainly using the checkmark to indicate selection rather than the background color:

Screenshot 2021-03-31 at 09 51 30

**Comment by Addison-Stavlo** [comment]
> I made a couple of adjustments to the design to better match other similar UI elements. Mainly using the checkmark to indicate selection rather than the background color:

Interesting, where else do we use this checkmark currently? (still waking up / drawing a blank on this 😆 )

>  to better match other similar UI elements.

This is what the button styles currently do by default (the opacity drop):

Screen Shot 2021-03-31 at 11 12 33 AM

It feels a bit ugly in this case to me.  I think the checkbox or dark background both work better.

**Comment by Addison-Stavlo** [comment]
> I made a couple of adjustments to the design

Oh, I just realized you added those changes to this PR. 😆  Thanks a bunch!

**Comment by Addison-Stavlo** [comment]
Here is a pic of me hovering over the selected button:

Screen Shot 2021-03-31 at 11 22 22 AM

I assume we want to get rid of the hover effect on the selected item (no change to blue)?

**Comment by jameskoster** [comment]
> where else do we use this checkmark currently? 

It's a slightly different application to this, but the checkmark appears in the options here:

Screenshot 2021-03-31 at 16 33 40

> I assume we want to get rid of the hover effect on the selected item (no change to blue)?

Ah yes, good spot.

**Comment by Addison-Stavlo** [comment]
>  the checkmark appears in the options here

Ah right!  Thanks for pointing it out.  I have seen them plenty of times in the past but they don't stick in my memory.  Thats probably a good sign that they do their job well without forcing me to think about them 😆

**Comment by Addison-Stavlo** [comment]
Updated a bit to see if we could get better a11y navigation that will scale with more options in the future.  I kind of got in a rabbit-hole with VoiceOver, things seem to work pretty well except my one complaint now is it doesn't announce when I'm in the "Area" selection region, it just immediately starts talking about the "General" option (while I think a voiceover user would be asking "what is this an option for?"). 🤔

**Comment by jameskoster** [comment]
I'm not too familiar with voice over functionality, do you think some helper text beneath the area options could help there? Or even simpler – a (slightly) more descriptive label?

**Comment by Addison-Stavlo** [comment]
> I'm not too familiar with voice over functionality, do you think some helper text beneath the area options could help there? Or even simpler – a (slightly) more descriptive label?

Im not super familiar with it as well.  But it seems like the label and/or helper text I tried aren't read when we navigate into that selection area.  (For the 'name' input above it is, and for our similar dropdown/select input for area in the block inspector for the TP block it is.. but I guess this interaction is a bit different for it).  It could be a quirk of the specific voiceover software as well.

All that said though, I think it is still usable and it may not be a huge deal (although it would be nice to fix that tiny bit).  It does announce that it is a radio type item selection, whether or not the item is selected, and reads the area description for the first item which does add a lot of context.

File: packages/edit-site/src/components/template-part-converter/convert-to-template-part.js
Comment (jeyip) [inline packages/edit-site/src/components/template-part-converter/convert-to-template-part.js]: I wonder if it would be worthwhile to consolidate `AREA_OPTIONS` with the template part variations updates we made. Some of the metadata looks similar.

https://github.com/WordPress/gutenberg/blob/2c491ad941060b5076bc43093ae7f81788d5b848/packages/block-library/src/template-part/variations.js#L9-L30

Comment (Addison-Stavlo) [reply packages/edit-site/src/components/template-part-converter/convert-to-template-part.js]: Great observation!

I actually just pushed some commits that define these descriptions in the php files where we defined the default template part areas.  That is then added to settings which is loaded into the core/editor store, where a selector maps the proper icons to each area.  The new selector `__experimentalGetTemplatePartAreas` in the `core/editor` store could then be used in these other places like you mention so things are consolidated and consistent (block variations, block inspector 'area' selection options, etc.)

I would prefer holding off connecting that selector to these other sections for a follow up though. 😁

**Comment by jeyip** [comment]
> Updated a bit to see if we could get better a11y navigation that will scale with more options in the future. I kind of got in a rabbit-hole with VoiceOver, things seem to work pretty well except my one complaint now is it doesn't announce when I'm in the "Area" selection region, it just immediately starts talking about the "General" option (while I think a voiceover user would be asking "what is this an option for?"). 🤔

I can reproduce this as well. I'm not entirely certain what's causing it, but I started digging into the wrapper BaseControl element we're using. A few things I noticed:
- Looking at how the <BaseControl /> component is structured, I'm fairly certain that `BaseControl` should only be wrapping _one_ form control element, and not multiple children.
- The `id` prop we pass to `BaseControl` is applied as a `for` attribute to the`BaseControl` label. We should be applying the same `id` to whatever form control element `BaseControl` is wrapping : [Link](https://developer.wordpress.org/block-editor/reference-guides/components/base-control/#usage)

```
    <BaseControl
        id="textarea-1"
        label="Text"
        help="Enter some text"
    >
        <textarea
            id="textarea-1"
        />
    </BaseControl>
```

**Comment by Addison-Stavlo** [comment]
> BaseControl should only be wrapping one form control element, and not multiple children.

The one element would be the `Composite` here, then the compsite's children` CompositeItem` are the options for that input.

>We should be applying the same id to whatever form control element BaseControl is wrapping

oh good catch!  I missed that same id was applied to the inner element.  I was thinking that id was set on the bascontrol and the for on its corresponding `label` element.  I tried it out, but it still seems to have the same behavior 🤔

File: packages/edit-site/src/components/template-part-converter/convert-to-template-part.js
Comment (jeyip) [reply packages/edit-site/src/components/template-part-converter/convert-to-template-part.js]: > I would prefer holding off connecting that selector to these other sections for a follow up though. 😁

Yep! Sounds good

**Comment by jeyip** [comment]
> The one element would be the Composite here, then the compsite's children CompositeItem are the options for that input.

Ahhh that makes sense. To clarify, did you add the `id` to the `Composite` component? 

> I tried it out, but it still seems to have the same behavior 🤔

I think it still might be an issue because I'm not sure if the `label` element generated by `BaseControl` and the label's `for` attribute can associate itself with a div ( the `Composite` component renders as a div ). I believe it can only bind to other control elements like `textarea` or `input`. Looking up more information...

**Edit:**
I found this excerpt from the whatwg html spec : [Link](https://html.spec.whatwg.org/multipage/forms.html#category-label)

"Some elements, not all of them form-associated, are categorized as labelable elements. These are elements that can be associated with a label element. `button`, `input` (if the type attribute is not in the Hidden state) `meter`, `output`, `progress`, `select`, `textarea`, form-associated custom elements"

I don't think the `label` element that's auto-generated from `BaseControl` can bind to the `Composite` component correctly, because `Composite` renders as a div and not as a labelable element.

**Comment by Addison-Stavlo** [comment]
🤔  That makes sense, but im not sure what to do about it.  We could use the 'as' prop if there was a form associated element that would work for this case.  This also seems to be a problem with some pre-existing components as well (such as RadioGroup/Radio), as the RadioGroup also renders as a div.

Note - I could be making an issue out of nothing.  Im not super familiar with using voiceover and after testing out other examples (like radio inputs in w3 etc.) the voiceover seems to read about the same as here.  Its probably best not to invest too much time into this issue before we get some feedback from a11y.

**Comment by Addison-Stavlo** [comment]
Besides that a11y ❓  in so much discussion above, this should be ready for review.

**Comment by david-szabo97** [comment]
a11y seems fine: 
image

Note how it says "Area, group" at the end.

File: phpunit/class-block-templates-test.php
Comment (david-szabo97) [inline phpunit/class-block-templates-test.php]: Can you explain the change here?

**Review by david-szabo97 — COMMENTED** [review]
One nitpick, otherwise it looks good

**Comment by azhiya** [comment]
This was discussed in the [accessibility meeting today ](https://wordpress.slack.com/archives/C02RP4X03/p1617380664399200)and the team asked for testing instructions. Is this something that can be part of the PR request, for example, 'How to Test'?

**Comment by Addison-Stavlo** [comment]
Hi @azhiya, thanks for taking a look!

>  Is this something that can be part of the PR request, for example, 'How to Test'?

Yes!  My apologies, adding testing instructions completely slipped my mind when I moved this out of draft mode.  They should be there now.

File: phpunit/class-block-templates-test.php
Comment (Addison-Stavlo) [reply phpunit/class-block-templates-test.php]: Oh thanks for catching that, that wasn't supposed to happen!  (updated)

I was testing out something unrelated to this PR and it must have snuck its way in to a commit - someone had suggested that perhaps both the defined constant and template->area were coming in as `undefined` in the php unit tests. 🤣

**Review by jameskoster — APPROVED** [review]
Approving design, pending a11y input.

**Comment by david-szabo97** [comment]
Can we try `RadioGroup`/`Radio` [components](https://developer.wordpress.org/block-editor/reference-guides/components/radio-group/) here? 😄 That seems to fit here and we would get proper battle-tested a11y with them.

**Comment by Addison-Stavlo** [comment]
> Can we try RadioGroup/Radio components here? 😄 That seems to fit here and we would get proper battle-tested a11y with them.

I did try them at some point last week.  They seemed to behave about the same as the `Composite`/`CompositeItem` in this case except required a few more style overrides to get the desired design.

**Comment by Addison-Stavlo** [comment]
> Can we try RadioGroup/Radio components here?

Update to try this out: I think it definitely works better as RadioGroup in the end!  The styles should also now be the same as before, with some minor updates to prevent 'jumpyness' on smaller screen sizes.

**Comment by Addison-Stavlo** [comment]
> Note how it says "Area, group" at the end.

Back to this comment from a few days ago.  I noticed on my end that wasn't the case: the dialogue stopped right before that after "selected, radio".  Now with the RadioGroup, it also says "selected, radio, 1 of 3" but doesn't say "Area" as well.  I wonder if it still does on your end? @david-szabo97

File: lib/full-site-editing/template-parts.php
Comment (ntsekouras) [inline lib/full-site-editing/template-parts.php]: I believe we should we add the `gutenberg` domain to `__()` calls.

File: lib/full-site-editing/template-parts.php
Comment (ntsekouras) [inline lib/full-site-editing/template-parts.php:170]: What happens to `WP_TEMPLATE_PART_AREA_SIDEBAR`? I know it's not been included to template part block variations. Is still valid but not exposed yet? Because it has code references like here: https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/navigation-sidebar/navigation-panel/constants.js#L94

File: packages/editor/src/store/selectors.js
Comment (ntsekouras) [inline packages/editor/src/store/selectors.js]: This is not a selector, but a `util`. Can we move this to another file?

File: packages/editor/src/store/selectors.js
Comment (ntsekouras) [inline packages/editor/src/store/selectors.js]: Nitpick: you don't need to declare `areasWithIcons`, return the result from `areas?.map`.

**Review by ntsekouras — COMMENTED** [review]
Thanks for working on this! I have left some comments but in general this looks good 👍

**Comment by david-szabo97** [comment]
> Back to this comment from a few days ago. I noticed on my end that wasn't the case: the dialogue stopped right before that after "selected, radio". Now with the RadioGroup, it also says "selected, radio, 1 of 3" but doesn't say "Area" as well. I wonder if it still does on your end?

Works fine for me 😬  It doesn't stop after "1 of 3", it goes through the end.

File: lib/full-site-editing/template-parts.php
Comment (Addison-Stavlo) [reply lib/full-site-editing/template-parts.php:170]: > What happens to WP_TEMPLATE_PART_AREA_SIDEBAR?

Basically getting rid of its definition here for now since it isn't yet supported in the editor.  Originally we were just defining and supporting header, footer, and general here - but I defined sidebar here knowing it would *eventually* be supported in the editor and front-end, and thought there would be no harm in declaring it as a valid area on the back end in the meantime since its been in all the design discussions.  However, since we are refactoring these definitions to be in-sync with the editor (send them to settings, store, .map them as options in inputs, etc.) it is now more important that they are 1-1 with what the editor supports.  Once we are ready to support sidebars in the editor, we would add their definition here again.

File: packages/editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/editor/src/store/selectors.js]: updated

File: packages/editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/editor/src/store/selectors.js]: yup, updated!

File: lib/full-site-editing/template-parts.php
Comment (Addison-Stavlo) [reply lib/full-site-editing/template-parts.php]: ah right! thanks.  updated 😁

**Comment by Addison-Stavlo** [comment]
Thanks for taking a look @ntsekouras !  Updated per your suggestions.

**Review by ntsekouras — APPROVED** [review]
Nice work @Addison-Stavlo 💯 

With happy CI (some valid php linting issues right now) let's 🚢 !

**Comment by paaljoachim** [comment]
Testing with WordPress 5.7
TT1
Special Gutenberg build version.

Selecting an image in the Index template.
Screen Shot 2021-04-12 at 23 04 07

Clicking "Make template part"

Screen Shot 2021-04-12 at 23 04 23

I see the screen with three options. General, Header and Footer.
To me General could be any sidebar area. Such as a template area just seen on posts. A under header area. A left or right sidebar. Or any other sidebar added to the site.

It uses the text:
General
General template often perform a specific role like displaying post content, and are not tied to any particular area.
---> To me it feels like it is missing some information which says that any area that are not associated with Header or Footer could be a general area.

**Comment by Addison-Stavlo** [comment]
Thanks for taking a look @paaljoachim!

> To me General could be any sidebar area. 
> To me it feels like it is missing some information which says that any area that are not associated with Header or Footer could be a general area.

We can definitely update the descriptions if needed.  But I think the idea in the future is to actually support a "Sidebar" area in addition to header/footer/general.  Its a bit more complicated than the other cases and we dont yet have the in-editor tools to support it, but I would assume once we get to developing sidebar support it would be its own item in this list? cc @jameskoster

**Comment by jameskoster** [comment]
We can perhaps simplify the description a bit by leaving out the example.

> General template parts are not tied to any particular area, and can appear in several locations across a site

Is that better?

**Comment by paaljoachim** [comment]
Yeah! That fits in nicely with a general template parts area.

**Comment by Addison-Stavlo** [comment]
https://github.com/WordPress/gutenberg/pull/30811 - heres a PR to update the description.  Feel free to update the copy if needed.

**Pull request merged**
